File: TicTacToe/TicTacToe_Node/components/board.ts
import { PlayingPiece } from "./playingPieces/playingPiece.ts";

export class Board {
  private _board: PlayingPiece[][];
  private _size: number;
  private _occupiedCount: number;

  constructor(size: number) {
    this._size = size;
    this._occupiedCount = 0;
    this._board = Array.from({ length: size }, () => Array(size).fill(null));
  }

  areCellsAvailable() {
    return this._occupiedCount < this._size * this._size;
  }

  makeMove(row: number, col: number, piece: PlayingPiece) {
    if (
      row < 0 ||
      row >= this._size ||
      col < 0 ||
      col >= this._size ||
      this._board[row][col] != null
    ) {
      return false;
    }

    this._board[row][col] = piece;
    this._occupiedCount++;
    return true;
  }

  hasWon() {
    // check horizontally
    for (let row = 0; row < this._size; row++) {
      for (let col = 1; col < this._size - 1; col++) {
        if (
          this._board[row][col - 1] !== null &&
          this._board[row][col] !== null &&
          this._board[row][col + 1] !== null &&
          this._board[row][col - 1].symbol === this._board[row][col].symbol &&
          this._board[row][col].symbol === this._board[row][col + 1].symbol
        ) {
          return true;
        }
      }
    }

    // check vertically
    for (let col = 0; col < this._size; col++) {
      for (let row = 1; row < this._size - 1; row++) {
        if (
          this._board[row - 1][col] !== null &&
          this._board[row][col] !== null &&
          this._board[row + 1][col] !== null &&
          this._board[row - 1][col].symbol === this._board[row][col].symbol &&
          this._board[row][col].symbol === this._board[row + 1][col].symbol
        ) {
          return true;
        }
      }
    }

    // check diagonally
    for (let i = 1; i < this._size - 1; i++) {
      if (
        this._board[i - 1][i - 1] !== null &&
        this._board[i][i] !== null &&
        this._board[i + 1][i + 1] !== null &&
        this._board[i - 1][i - 1].symbol === this._board[i][i].symbol &&
        this._board[i][i].symbol === this._board[i + 1][i + 1].symbol
      ) {
        return true;
      }
    }

    // check anti-diagonally
    for (let i = 1; i < this._size - 1; i++) {
      const n = this._size - 1;
      if (
        this._board[n - (i - 1)][i - 1] !== null &&
        this._board[n - i][i] !== null &&
        this._board[n - (i + 1)][i + 1] !== null &&
        this._board[n - (i - 1)][i - 1].symbol ===
          this._board[n - i][i].symbol &&
        this._board[n - i][i].symbol === this._board[n - (i + 1)][i + 1].symbol
      ) {
        return true;
      }
    }

    return false;
  }

  displayBoard() {
    for (let i = 0; i < this._size; i++) {
      for (let j = 0; j < this._size; j++) {
        if (this._board[i][j] === null) {
          process.stdout.write("null");
        } else {
          process.stdout.write(`${this._board[i][j].symbol}`);
        }
        if (j < this._size - 1) process.stdout.write(` | `);
      }
      console.log("");
    }
  }
}

File: TicTacToe/TicTacToe_Go/components/board.go
package components

import (
	"fmt"
	"ticTacToe/components/playingPieces"
)

func GetNewBoard(size int) *Board {
	board := make([][]playingPieces.IPlayingPiece, size)
	for i := 0; i < size; i++ {
		board[i] = make([]playingPieces.IPlayingPiece, size)
	}

	return &Board{
		board:         board,
		size:          size,
		occupiedCount: 0,
	}
}

type Board struct {
	board         [][]playingPieces.IPlayingPiece
	size          int
	occupiedCount int
}

func (b *Board) MakeMove(row, col int, piece playingPieces.IPlayingPiece) bool {
	if row < 0 || row >= b.size || col < 0 || col >= b.size ||
		!b.AreCellsAvailable() || b.board[row][col] != nil {

		return false
	}
	b.board[row][col] = piece
	b.occupiedCount++
	return true
}

func (b *Board) AreCellsAvailable() bool {
	return b.occupiedCount < (b.size * b.size)
}

func (b *Board) HasWon() bool {
	// check horizontally
	for i := 0; i < b.size; i++ {
		for j := 1; j < b.size-1; j++ {
			if b.board[i][j-1] != nil && b.board[i][j] != nil && b.board[i][j+1] != nil &&
				b.board[i][j-1].GetSymbol() == b.board[i][j].GetSymbol() &&
				b.board[i][j].GetSymbol() == b.board[i][j+1].GetSymbol() {
				return true
			}
		}
	}

	// check vertically
	for i := 0; i < b.size; i++ {
		for j := 1; j < b.size-1; j++ {
			if b.board[j-1][i] != nil && b.board[j][i] != nil && b.board[j+1][i] != nil &&
				b.board[j-1][i].GetSymbol() == b.board[j][i].GetSymbol() &&
				b.board[j][i].GetSymbol() == b.board[j+1][i].GetSymbol() {
				return true
			}
		}
	}

	// check diagonal
	for i := 1; i < b.size-1; i++ {
		if b.board[i-1][i-1] != nil && b.board[i][i] != nil && b.board[i+1][i+1] != nil &&
			b.board[i-1][i-1].GetSymbol() == b.board[i][i].GetSymbol() &&
			b.board[i][i].GetSymbol() == b.board[i+1][i+1].GetSymbol() {

			return true
		}
	}

	// check anti-diagonal
	for i := 1; i < b.size-1; i++ {
		if b.board[b.size-(i-1)-1][i-1] != nil && b.board[b.size-i-1][i] != nil && b.board[b.size-(i+1)-1][i+1] != nil &&
			b.board[b.size-(i-1)-1][i-1].GetSymbol() == b.board[b.size-i-1][i].GetSymbol() &&
			b.board[b.size-i-1][i].GetSymbol() == b.board[b.size-(i+1)-1][i+1].GetSymbol() {

			return true
		}
	}
	return false
}

func (b *Board) DisplayBoard() {
	for i := 0; i < b.size; i++ {
		for j := 0; j < b.size; j++ {
			if b.board[i][j] == nil {
				fmt.Print("nil")
			} else {
				fmt.Print(b.board[i][j].GetSymbol())
			}

			if j < b.size-1 {
				fmt.Print(" | ")
			}
		}
		fmt.Println()
	}
}
